libkbfs/block_retrieval_queue.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"container/heap"
+	"sync"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultBlockRetrievalWorkerQueueSize int = 100
+	defaultOnDemandRequestPriority       int = 100
+)
+
+// blockRetrievalRequest represents one consumer's request for a block.
+type blockRetrievalRequest struct {
+	ctx    context.Context
+	block  Block
+	doneCh chan error
+}
+
+// blockRetrieval contains the metadata for a given block retrieval. May
+// represent many requests, all of which will be handled at once.
+type blockRetrieval struct {
+	//// Retrieval Metadata
+	// the block pointer to retrieve
+	blockPtr BlockPointer
+	// the key metadata for the request
+	kmd KeyMetadata
+	// the individual requests for this block pointer: they must be notified
+	// once the block is returned
+	requests []*blockRetrievalRequest
+
+	//// Queueing Metadata
+	// the index of the retrieval in the heap
+	index int
+	// the priority of the retrieval: larger priorities are processed first
+	priority int
+	// state of global request counter when this retrieval was created;
+	// maintains FIFO
+	insertionOrder uint64
+}
+
+// blockRetrievalQueue manages block retrieval requests. Higher priority
+// requests are executed first. Requests are executed in FIFO order within a
+// given priority level.
+type blockRetrievalQueue struct {
+	mtx sync.RWMutex
```

Comment:
can you add a comment re. what mtx protects? looks like everything but `workerQueue`?